### PR TITLE
alpha 5

### PR DESCRIPTION
**Pull request by JustFly1984**
added new eslint plugin, added optional onUnmount prop for each component. improved types, get rid of prettier.

### CODE REVIEW

File: packages/react-google-maps-api/.eslintrc.js [hunk 1]
```diff
@@ -33,8 +34,8 @@ module.exports = {
     'standard',
     'no-inferred-method-name',
     'react-functional-set-state',
-    '@typescript-eslint',
-    'prettier'
+    '@typescript-eslint'
+    // 'prettier'
```
Comment (uriklar) [inline packages/react-google-maps-api/.eslintrc.js]: I'll open an issue regarding prettier and linting in general because I see we're spending too much of our time over conflicting rules and setups

File: packages/react-google-maps-api/.eslintrc.js [hunk 2]
```diff
@@ -63,6 +64,27 @@ module.exports = {
     }
   },
   rules: {
+    '@getify/proper-arrows/params': [
+      'error', {
+        'unused': 'trailing'
+      }
+    ],
+    '@getify/proper-arrows/name': [
```
Comment (uriklar) [inline packages/react-google-maps-api/.eslintrc.js]: Is this one really neccesary? or the return one?

> "name": requires => arrow functions to only be used in positions where they receive an inferred name (i.e., assigned to a variable or property, etc), to avoid the poor readbility/debuggability of anonymous function expressions.

I have never seen a codebase that gives names to anonymous arrow functions...

File: packages/react-google-maps-api/src/GoogleMap.tsx [hunk 3]
```diff
@@ -65,101 +65,102 @@ const updaterMap = {
...
+  onZoomChanged?: () => void;
+  onLoad?: (map: google.maps.Map) => void | Promise<void>;
+  onUnmount?: (map: google.maps.Map) => void | Promise<void>;
 }
 
 export class GoogleMap extends React.PureComponent<
   GoogleMapProps,
   GoogleMapState
 > {
-  public static defaultProps: GoogleMapProps = {
```
Comment (uriklar) [inline packages/react-google-maps-api/src/GoogleMap.tsx]: why remove default props?

File: packages/react-google-maps-api/src/utils/helper.ts [hunk 4]
```diff
@@ -6,8 +6,9 @@ const applyUpdaterToNextProps = (
   updaterMap: any,
   prevProps: any,
   nextProps: any,
+  // eslint-disable-next-line @getify/proper-arrows/params
```
Comment (uriklar) [inline packages/react-google-maps-api/src/utils/helper.ts]: first of all I think we should remove the params length rule. 
second, easiest way to solve this error is to make this a regular function as it should be

File: packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx [hunk 5]
```diff
@@ -128,51 +131,72 @@ export class MarkerClustererComponent extends PureComponent<
     markerClusterer: null
   }
 
-  componentDidMount = () => {
+  // eslint-disable-next-line @getify/proper-arrows/this, @getify/proper-arrows/name
+  setMarkerClustererCallback = (): void => {
```
Comment (uriklar) [inline packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: We should avoid using class methods as arrow function. 
See: https://medium.com/@rohans16996/dark-side-of-the-fat-arrow-function-8d2a477f743b

TL;DR - when using arrow function as class properties the function get's duplicated on each component instance. Usually this is negligible, but in our lib you might be rendering out thousands of components so this starts to add up

File: packages/react-google-maps-api/.eslintrc.js (hunk 1 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/.eslintrc.js]: I found that I can wrap code instead of horizontal scrolling, I can live with it.

File: packages/react-google-maps-api/.eslintrc.js (hunk 2 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/.eslintrc.js]: well, I think at some point we will be moving to hooks, and there will be a lot of anonymous functions. I already have all Components refactored to functional in my other project, so I think it is safe to make functional version, and compare bundles on size and performance.

File: packages/react-google-maps-api/src/utils/helper.ts (hunk 4 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/utils/helper.ts]: I agree, but I'm not sure how. As I understand, this eslint plugin evolves, and this rules a bit more complex. By now this comments absolutely fine.

File: packages/react-google-maps-api/src/GoogleMap.tsx (hunk 3 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/GoogleMap.tsx]: first of all, if prop is undefined, it is marked as `propName?: string; in interface declaration.

in this case we do not need default props, especially noop functions.

File: packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx (hunk 5 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: >>>
It`s generally advised to keep *bound* functions out of React Components.
In this case, every single time the component instance is created, a new bound function will be created. which in turn slows down React rendering! Hence, AVOID!

Nothing says about arrow functions in React components, only bound.

Comment (JustFly1984) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: React Components lifecycle methods does not required to be bound, but all other methods are required. Since React async, we are generally avoiding constructors, so the only one way is to keep all instance methods as arrow functions, if we use this in function scope. Otherwise we can just extract that function outside the module to `./src/utils/` or elsewhere.

File: packages/react-google-maps-api/src/GoogleMap.tsx (hunk 3 above)
Comment (uriklar) [reply packages/react-google-maps-api/src/GoogleMap.tsx]: I agree it's just a matter of taste weather to use defaultProps with noop or a condition checking if it is defined. But just FYI the way Typescript handles default props is that you remove the question mark and the prop is still optional because it has a default. I tested thia behavior and it works. But again, it works this way as well

File: packages/react-google-maps-api/.vscode/settings.json [hunk 6]
```diff
@@ -7,8 +7,8 @@
   "eslint.enable": true,
   "eslint.autoFixOnSave": true,
   "eslint.alwaysShowStatus": true,
-  "eslint.run": "onSave",
-  "eslint.trace.server": "messages",
+  "eslint.run": "onType",
```
[the remark this answers is not in the document]
Comment (JustFly1984) [reply packages/react-google-maps-api/.vscode/settings.json]: it is fine to have "onType" then whole project is typed, and you do new stuff, but I had experienced issues with vscode performance, then I was working on refactoring huge JS codebase to TS, that is why I switched it "onSave".

File: packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx (hunk 5 above)
Comment (uriklar) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: Why should we avoid constructors? The only issue about constructor function is that it shouldn't contain side effects. Other than that it's fine.
As you say life cycle methods are automatically bound, but any other function (using this) has to be bound. Using the arrow function syntax compiles to defining the function per instance as the article says

Comment (uriklar) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: Why should we avoid constructors? The only issue about constructor function is that it shouldn't contain side effects. Other than that it's fine.
As you say life cycle methods are automatically bound, but any other function (using this) has to be bound. Using the arrow function syntax compiles to defining the function per instance as the article says

File: packages/react-google-maps-api/.eslintrc.js (hunk 2 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/.eslintrc.js]: That is for debuggability - The guy who made this eslint plugin knows what he is doing, but there is no react support out of the box. Maybe we can disable that rule. Well, the thing is that rule asks to refactor () => {} to function functionName() {}

Comment (JustFly1984) [reply packages/react-google-maps-api/.eslintrc.js]: In general this rule will help less advanced developers to read the code too.

Comment (uriklar) [reply packages/react-google-maps-api/.eslintrc.js]: Yeah i read the motivation behind the rule 😊 it just feels a bit harsh... I've never seen code that gives names to anonymous function. I have no principal objection to it though

File: packages/react-google-maps-api/src/GoogleMap.tsx (hunk 3 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/GoogleMap.tsx]: I prefer to be consistent) The only one issue which we got, where we could not get rid default props, is functional components, which requires some props default value to be an object or array. If you so default objects or array in props destructuring, each time those props has new object or array, which causing component to re-render. And defaultProps helps with avoiding this.

File: packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx (hunk 5 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: but bound functions also goes into the object, so there is no gain)

Comment (JustFly1984) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: everything in constructor could be called multiple times, so better to do not run constructor at all in Components. Soon we will migrate to hooks, and there is no constructors.

in react component all functions goes into an object, except if you are extending some other class, which extends Component or PureComponent.

all lifecycle methods are bound by react, all custom function methods should be bound in constructor. bounding a function returns a copy of the function.

Arrow class methods don't have to be bound in babel environment cos we have a proposal babel plugin to handle that case.

File: packages/react-google-maps-api/src/utils/helper.ts (hunk 4 above)
Comment (uriklar) [reply packages/react-google-maps-api/src/utils/helper.ts]:
```
function applyUpdaterToNextProps(
  updaterMap: any,
  prevProps: any,
  nextProps: any,
  instance: any
): any {
  let map: any = {}

  const iter = (fn: any, key: string) => {
    const nextValue = nextProps[key]

    if (nextValue !== prevProps[key]) {
      map[key] = nextValue
      fn(instance, nextValue)
    }
  }

  forEach(updaterMap, iter)

  return map
}
```

this removes the error

File: packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx (hunk 5 above)
Comment (uriklar) [reply packages/react-google-maps-api/src/components/addons/MarkerClusterer.tsx]: You are absolutely... right! I just read this: https://github.com/facebook/react/issues/9851
Ok so that's ok then.

**Comment by uriklar** [comment]
Issues from testing:
[ ] Docs - Polygon/Polyline/HeatmapLayer examples are broken
[ ] Gatsby - loading script and then opening Data example makes app crash
